collection/collection/build.gradle

```diff
@@ -64,16 +67,37 @@ kotlin {
                 dependsOn commonMain
             }
 
-            configure([linuxX64Main, macosX64Main, macosArm64Main]) {
-                dependsOn nativeMain
-            }
-
             nativeTest {
                 dependsOn commonTest
             }
 
-            configure([linuxX64Test, macosX64Test, macosArm64Test]) {
-                dependsOn nativeTest
+            darwinMain {
+                dependsOn(nativeMain)
+            }
+            linuxMain {
+                dependsOn(nativeMain)
+            }
+
+            targets.all { target ->
+                if (target.platformType == KotlinPlatformType.native) {
+                    target.compilations["main"].defaultSourceSet {
+                        if (target.konanTarget.family == Family.OSX) {
+                            dependsOn(darwinMain)
+                        } else if (target.konanTarget.family == Family.LINUX) {
+                            dependsOn(linuxMain)
+                        } else {
+                            throw new GradleException("unknown native target ${target}")
+                        }
+                    }
+                    target.compilations["test"].defaultSourceSet {
+                        dependsOn(nativeTest)
+                    }
+                }
+            }
+            // Workaround for https://youtrack.jetbrains.com/issue/KT-51763
+            // Make sure commonization runs before any compilation task.
+            tasks.withType(KotlinNativeCompile).configureEach {
```

Comment:
Do we have plans / bug to track moving this into AndroidX plugins so we don't have to copy this to every project?

Reply:
yea, that is the sourceset work that @dsaff is looking into.
I also already have the prototype: https://android-review.googlesource.com/c/platform/frameworks/support/+/2072820

waiting on @dsaff to have some time to play with it then we should be able to move forward.